pkg/client/client.go

```diff
@@ -124,19 +124,15 @@ func newClient(config *rest.Config, options Options) (*client, error) {
 		config.UserAgent = rest.DefaultKubernetesUserAgent()
 	}
 
-	if !options.WarningHandler.SuppressWarnings {
-		// surface warnings
-		logger := log.Log.WithName("KubeAPIWarningLogger")
-		// Set a WarningHandler, the default WarningHandler
-		// is log.KubeAPIWarningLogger with deduplication enabled.
-		// See log.KubeAPIWarningLoggerOptions for considerations
-		// regarding deduplication.
-		config.WarningHandler = log.NewKubeAPIWarningLogger(
-			logger,
-			log.KubeAPIWarningLoggerOptions{
-				Deduplicate: !options.WarningHandler.AllowDuplicateLogs,
-			},
-		)
+	// By default, we de-duplicate and surface warnings.
+	config.WarningHandler = log.NewKubeAPIWarningLogger(
```

Comment:
Good question. I considered this, and actually prefer it.

I thought it might be seen as a backward-incompatible change, so I did not include it in this PR.

However, thinking about it now, I realize that, by default, config.WarningHandler is nil. So, in the default case, we would continue to modify warning handler, but we would also respect the warning handler the user chose.

I'll make the change.

Reply:
On a related note, I thought it might be easier to remove `options.WarningHandler` altogether, and ask the user to define `config.WarningHandler`. That might annoy users.

Reply:
The more I think about that, the more I like it. How about: If you define `config.WarningHandler`, we'll respect it. If you don't , we'll use `KubeAPIWarningLogger` configured some default way.

If you don't like its default configuration, you can always choose to instantiate `KubeAPIWarningLogger` with a custom configuration, and yourself, and assign it to `config.WarningHandler`.

Reply:
This is the same approach we take with user agent:
https://github.com/kubernetes-sigs/controller-runtime/blob/16d25f117642cc6cfe4c5859bf8cc2a305843750/pkg/client/client.go#L123-L125

Reply:
> How about: If you define config.WarningHandler, we'll respect it. If you don't , we'll use KubeAPIWarningLogger configured some default way.

We'll have to deprecate `options.WarningHandler` first, and when we remove it, we can adjust the logic.

Reply:
I created https://github.com/dlipovetsky/controller-runtime/pull/1. Let me know what you think.

I think we can merge this PR, because it preserves existing behavior (which does not respect `config.WarningHandler` by default).

Then I can open a new PR (against this repo) from https://github.com/dlipovetsky/controller-runtime/pull/1

Reply:
My main goal is to be able to configure a custom warning handler. E.g. for Cluster API it could be useful to only drop certain warnings (the finalizer one, in controllers).

I'm not entirely sure if we should deprecate the WarningHandler field or not

Reply:
We could also allow folks configuring a WarningHandler in the WarningHandlerOptions

Reply:
Guess this depends overall if we think we should:
* just find a way that user can configure a WarningHandler
* also respect what we get from the rest config

Reply:
In https://github.com/kubernetes-sigs/controller-runtime/pull/2887#discussion_r1692085803 I demonstrated that we cannot meet both goals at the same time.

Working from the "principle of least suprise," I think it's better to deprecate WarningHandlerOptions, and ask the user to use the rest config.

Reply:
Fine for me! Let's see what others say on the new PR